src/gpuarray_reduction.c

```diff
@@ -6,841 +6,4095 @@
+	    reverseS0 = 0, reverseD0 = 0, reverseD1 = 0;
+	
+	if (axisIsReduced     (into) != axisIsReduced     (from)                 ||
+	    axisGetS0AbsStride(into) != axisGetS0AbsStride(from)*axisGetLen(from)){
+		return 0;
+	}
+	
+	if (reduxInvRequiresD0(ctx)  &&
+	    axisGetD0AbsStride(into) != axisGetD0AbsStride(from)*axisGetLen(from)){
+		return 0;
+	}
+	
+	if (reduxInvRequiresD1(ctx)  &&
+	    axisGetD1AbsStride(into) != axisGetD1AbsStride(from)*axisGetLen(from)){
+		return 0;
+	}
+	
+	signS0    = (axisGetS0Stride(into)^axisGetS0Stride(from)) < 0;
+	signD0    = (axisGetD0Stride(into)^axisGetD0Stride(from)) < 0;
+	signD1    = (axisGetD1Stride(into)^axisGetD1Stride(from)) < 0;
+	reverseS0 = signS0;
+	reverseD0 = signD0 && reduxInvRequiresD0(ctx);
+	reverseD1 = signD1 && reduxInvRequiresD1(ctx);
+	
+	if (GpuReductionAttr_issensitive(&ctx->gr->grAttr)){
+		if (reverseS0 || reverseD0 || reverseD1){
+			return 0;
+		}
+	}
+	
+	if (reduxInvRequiresD0(ctx) &&
+	    reduxInvRequiresD1(ctx) &&
+	    reverseD0 != reverseD1){
+		/* Either both, or neither, of dst and dstArg must require reversal. */
+		return 0;
+	}
+	
+	if (reverseS0){
+		ctx->S0Off += (ssize_t)(axisGetLen(from)-1)*axisGetS0Stride(from);
+		into->s0S   = -axisGetS0Stride(from);
+	}else{
+		into->s0S   =  axisGetS0Stride(from);
+	}
+	
+	if (reverseD0){
+		ctx->D0Off += (ssize_t)(axisGetLen(from)-1)*axisGetD0Stride(from);
+		into->d0S   = -axisGetD0Stride(from);
+	}else{
+		into->d0S   =  axisGetD0Stride(from);
+	}
+	
+	if (reverseD1){
+		ctx->D1Off += (ssize_t)(axisGetLen(from)-1)*axisGetD1Stride(from);
+		into->d1S   = -axisGetD1Stride(from);
+	}else{
+		into->d1S   =  axisGetD1Stride(from);
+	}
+	
+	into->len *= axisGetLen(from);
+	
+	return 1;
+}
+
+/**
+ * Sort an array of *pointers* to axes by the given comparison function, while
+ * not touching the axes themselves.
+ */
+
+static void        reduxSortAxisPtrsBy              (axis_desc**            ptrs,
+                                                     axis_desc*             axes,
+                                                     size_t                 numAxes,
+                                                     int(*fn)(const void*, const void*)){
+	size_t i;
+	
+	for (i=0;i<numAxes;i++){
+		ptrs[i] = &axes[i];
+	}
+	
+	qsort(ptrs, numAxes, sizeof(*ptrs), fn);
+}
+
+
+/**
+ * @brief Initialize generator context.
+ * 
+ * After this function, calling reduxGenCleanup*() becomes safe.
+ */
+
+static int         reduxGenInit                     (GpuReduction*        gr){
+	gr->kArgTypeCodes = NULL;
+	gr->kSourceCode   = NULL;
+	gr->kErrorString  = NULL;
+	gr->kNumArgs      = 0;
+	
+	return reduxGenInferProperties(gr);
+}
+
+/**
+ * @brief Begin inferring the properties of the reduction operator.
+ */
+
+static int         reduxGenInferProperties          (GpuReduction*        gr){
+	int i;
+	
+	/**
+	 * Source code buffer preallocation failed?
+	 */
+	
+	if (strb_ensure(&gr->s, 32*1024) != 0){
```

Comment:
I generate 8 different versions of the core loops depending on where the "split" and "huge" axes are, so the generated source code does get a bit voluminous. Could be made smaller with macros.